helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableNode.java

```diff
@@ -19,10 +19,293 @@
+  private static final Logger _logger = LoggerFactory.getLogger(AssignableNode.class.getName());
+
+  // basic node information
+  private final String _instanceName;
+  private Set<String> _instanceTags;
+  private String _faultZone;
+  private Map<String, List<String>> _disabledPartitionsMap;
+  private Map<String, Integer> _maxCapacity;
+  private int _maxPartition;
+
+  // proposed assignment tracking
+  // <resource name, partition name>
+  private Map<String, Set<String>> _currentAssignments;
+  // <resource name, top state partition name>
+  private Map<String, Set<String>> _currentTopStateAssignments;
+  // <capacity key, capacity value>
+  private Map<String, Integer> _currentCapacity;
+  // runtime usage tracking
+  private int _totalReplicaAssignmentCount;
+  private float _highestCapacityUtilization;
+
+  AssignableNode(ResourceControllerDataProvider clusterCache, String instanceName,
+      Collection<AssignableReplica> existingAssignment) {
+    _instanceName = instanceName;
+    refresh(clusterCache, existingAssignment);
+  }
+
+  private void reset() {
+    _currentAssignments = new HashMap<>();
+    _currentTopStateAssignments = new HashMap<>();
+    _currentCapacity = new HashMap<>();
+    _totalReplicaAssignmentCount = 0;
+    _highestCapacityUtilization = 0;
+  }
+
+  /**
+   * Update the node with a ClusterDataCache. This resets the current assignment and recalculate currentCapacity.
+   * NOTE: While this is required to be used in the constructor, this can also be used when the clusterCache needs to be
+   * refreshed. This is under the assumption that the capacity mappings of InstanceConfig and ResourceConfig could
+   * subject to changes. If the assumption is no longer true, this function should become private.
+   *
+   * @param clusterCache - the current cluster cache to initial the AssignableNode.
+   */
+  private void refresh(ResourceControllerDataProvider clusterCache,
+      Collection<AssignableReplica> existingAssignment) {
+    reset();
+
+    InstanceConfig instanceConfig = clusterCache.getInstanceConfigMap().get(_instanceName);
+    ClusterConfig clusterConfig = clusterCache.getClusterConfig();
+
+    _currentCapacity.putAll(instanceConfig.getInstanceCapacityMap());
+    _faultZone = computeFaultZone(clusterConfig, instanceConfig);
+    _instanceTags = new HashSet<>(instanceConfig.getTags());
+    _disabledPartitionsMap = instanceConfig.getDisabledPartitionsMap();
+    _maxCapacity = instanceConfig.getInstanceCapacityMap();
+    _maxPartition = clusterConfig.getMaxPartitionsPerInstance();
+
+    assignNewBatch(existingAssignment);
+  }
+
+  /**
+   * Assign a replica to the node.
+   *
+   * @param assignableReplica - the replica to be assigned
+   */
+  void assign(AssignableReplica assignableReplica) {
+    if (!addToAssignmentRecord(assignableReplica, _currentAssignments)) {
+      throw new HelixException(String
+          .format("Resource %s already has a replica from partition %s on this node",
```

Comment:
Can we add the node information as well? This assignment implementation is very similar to that in Task Framework, and I realized if we do not include the node name, a lot of times the logs become a lot less helpful.

Reply:
Sounds good.